mail-server/users.nix

```diff
@@ -22,36 +22,114 @@ let
   vmail_user = {
     name = vmailUserName;
     isNormalUser = false;
-    uid = vmailUIDStart;
+    uid = vmailUID;
     home = mailDirectory;
     createHome = true;
     group = vmailGroupName;
   };
 
   # accountsToUser :: String -> UserRecord
   accountsToUser = account: {
-    name = account.name;
-    isNormalUser = false;
-    group = vmailGroupName;
-    inherit (account) hashedPassword;
+    name = lib.head (lib.splitString "@" account.name);
+    domain = lib.head (lib.tail (lib.splitString "@" account.name));
+    inherit (account) hashedPassword sieveScript;
   };
 
-  # mail_users :: { [String]: UserRecord }
-  mail_users = lib.foldl (prev: next: prev // { "${next.name}" = next; }) {}
-    (map accountsToUser (lib.attrValues loginAccounts));
+  virtualMailUsersActivationScript = pkgs.writeScript "activate-virtual-mail-users" ''
+    #!${pkgs.stdenv.shell}
+
+    set -euo pipefail
+
+    # Create mailDirectory if it doesn't exist.
+    if (! test -d "${mailDirectory}"); then
+      mkdir -p "${mailDirectory}"
+      chown ${vmailUserName}:${vmailGroupName}
+      chmod 700 "${mailDirectory}"
+    fi
+
+    # Create a passwd and shadow file under mailDirectory/domain for each domain
+    # in cfg.domains.
+    ${lib.concatMapStringsSep "\n"
+      (domain: let domainDir = "${mailDirectory}/${domain}"; in ''
+        # Create per-domain directory if it doesn't exist.
+        if (! test -d "${domainDir}"); then
+          mkdir -p "${domainDir}"
```

Comment:
Do we need a `chown` here too? How do we allow `vmail` write access, via the group?

Reply:
That is taken care of here: https://github.com/r-raymond/nixos-mailserver/pull/37/files#diff-8916cc3e7ea3ee3d68abb278c05c8978R59

700 is sufficient because the virtual mail user remains the owner of the virtual user dirs.